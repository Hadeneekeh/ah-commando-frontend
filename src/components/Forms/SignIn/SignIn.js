import React, { Component } from 'react';
import { withRouter, Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import Loader from 'react-loader-spinner';
import Modal from '@Components/Modal';
import Input from '@Components/Input';
import Button from '@Components/Button';
import connectComponent from '@Lib/connect-component';
import { closeModal, openModal } from '@Actions/uiActions';
import { logIn } from '@Actions/authActions';
import { validate, emailSchema, passwordSchema } from '@Utils/';
import './SignIn.scss';

export class SignIn extends Component {
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      password: '',
      showPassword: false,
      errors: {},
      isFormValid: true,
    };
  }

  handleChange = (e) => {
    const { name, value } = e.target;
    const errors = {};
    const [errorValue] = validate({
      [name]: value,
    }, name === 'password' ? passwordSchema : emailSchema);

    errors[name] = errorValue || '';
    this.setFormValidity(errors);
    this.setState({
      [name]: value,
      errors,
    });
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const { signIn, history } = this.props;
    const {
      email,
      password,
      errors,
    } = this.state;
    const formIsValid = this.validateForm();

    if (!formIsValid) {
      return this.setFormValidity(errors);
    }
    signIn({ email, password }, history);
  }

  validateForm = () => {
    const errors = {};
    const {
      email,
      password,
    } = this.state;
    const [emailError] = validate({ email }, emailSchema);
    const [passwordError] = validate({ password }, passwordSchema);

    errors.email = emailError || '';
    errors.password = passwordError || '';
    return this.setFormValidity(errors);
  }

  setFormValidity = (errors) => {
    let valid = true;
    Object.values(errors).forEach((value) => {
      if (value.length > 0) {
        valid = false;
      }
    });
    this.setState({ isFormValid: valid });

    return valid;
  }

  toggleVisibility = () => {
    this.setState(prevState => ({
      showPassword: !prevState.showPassword,
    }));
  }

  resetModal = () => {
    const { close } = this.props;
    this.setState({
      email: '',
      password: '',
      showPassword: false,
      errors: {},
      isFormValid: true,
    });
    close();
  }

  render() {
    const {
      email,
      password,
      showPassword,
      errors,
      isFormValid,
    } = this.state;
    const {
      ui: {
        loading,
        modalOpen,
        modal,
      }, requestPassword,
    } = this.props;
    const loader = <Loader type="BallTriangle" color="#fff" height={18} width={79} />;

    return (
      <Modal data-test="signup-form" close={this.resetModal} open={modalOpen && modal === 'signin'}>
        <div className="form">
          <h3 className="form-header">Sign in</h3>
          <p className="form-text">
            Sign in to experience authorâ€™s haven great community of creative minds.
            Follow your favourite authors and interact with the articles you love
          </p>
          <form style={{ display: 'flex', flexDirection: 'column', justifyContent: 'center' }}>
            <Input
              name="email"
              value={email}
              type="email"
              handleChange={this.handleChange}
              placeholder="john.doe@foo.bar"
              label="Email"
              error={errors.email}
            />
            <Input
              name="password"
              value={password}
              type="password"
              handleChange={this.handleChange}
              placeholder="**********"
              label="Password"
              togglable
              visible={showPassword}
              handleToggle={this.toggleVisibility}
              error={errors.password}
            />
            <Button
              datatest="login-button"
              label={loading ? null : 'Sign In'}
              handleClick={this.handleSubmit}
              disabled={loading ? true : !isFormValid}
              type="submit"
              datatest="test"
              style={{
                height: '45px',
                width: '300px',
                color: '#ffc700',
                backgroundColor: '#000',
                borderRadius: '0',
              }}
            >
              {loading && loader}
            </Button>
          </form>
          <div className="alternative-login">
            <p>Or create an account Using:</p>
            <div className="social-login">
              <span>
                <i className="fab fa-google fa-lg" style={{ color: 'red' }} />
                Google
              </span>
              <span>
                <i className="fab fa-facebook fa-lg" style={{ color: 'blue' }} />
                Facebook
              </span>
            </div>
          </div>
          <div className="switch-context">
            <p>
              No account?
              {' '}
              <Link to="/">
                Sign up
              </Link>
            </p>
          </div>
          <div className="forgot-password">
            <Button
              label="Forgot Password?"
              handleClick={requestPassword}
              type="button"
              style={{
                backgroundColor: '#fff',
                color: '#8075e6',
                borderRadius: '0',
                cursor: 'pointer',
                fontSize: '12px',
                padding: '0px',
                margin: '0px',
              }}
            />
          </div>
        </div>
      </Modal>
    );
  }
}

SignIn.propTypes = {
  ui: PropTypes.shape({
    loading: PropTypes.bool.isRequired,
    modalOpen: PropTypes.bool.isRequired,
    modal: PropTypes.string.isRequired,
  }).isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
  close: PropTypes.func.isRequired,
  signIn: PropTypes.func.isRequired,
  requestPassword: PropTypes.func.isRequired,
};

export default connectComponent(
  withRouter(SignIn), {
    close: closeModal,
    signIn: logIn,
    requestPassword: () => openModal('passwordModal'),
  },
);
