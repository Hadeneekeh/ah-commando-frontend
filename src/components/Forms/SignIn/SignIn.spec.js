import React from 'react';
import chai from 'chai';
import { shallow } from 'enzyme';
import { SignIn } from '@Components/Forms/SignIn/SignIn';

chai.should();

describe('<SignIn /> Component', () => {
  let wrapper, props, formHeader, instance, state;

  beforeEach(() => {
    state = {
      email: 'john.doe@testmail.com',
      password: 'P@ssword123',
      showPassword: false,
      errors: {},
      isFormValid: true,
    };
    props = {
      ui: {
        loading: false,
        modalOpen: true,
        modal: 'signin',
      },
      history: {
        push: jest.fn(),
      },
      close: jest.fn(),
      signIn: jest.fn(),
      signinViaSocial: jest.fn(),
    };
    wrapper = shallow(<SignIn {...props} />);
    wrapper.setState(state);
    instance = wrapper.instance();
    formHeader = wrapper.find('h3');
  });

  it('Renders as expected', () => {
    formHeader.text().should.equal('Sign in');
    wrapper.containsMatchingElement(<h3>Sign in</h3>).should.equal(true);
    wrapper.containsMatchingElement(
      <p>
        Sign in to experience authorâ€™s haven great community of creative minds.
        Follow your favourite authors and interact with the articles you love
      </p>,
    ).should.equal(true);
    wrapper.state().should.deep.equal(state);
  });

  it('Should close the modal when the resetModal function is called', () => {
    instance.resetModal();

    expect(props.close).toHaveBeenCalled();
  });

  it('Should call the singIn function when the form is submitted', () => {
    const event = {
      preventDefault: jest.fn(),
    };
    instance.handleSubmit(event);
    expect(props.signIn).toHaveBeenCalled();
  });

  it('Should not submit the form if it\'s invalid', () => {
    const event = {
      preventDefault: jest.fn(),
    };
    instance.setFormValidity = jest.fn();

    instance.validateForm = jest.fn().mockReturnValueOnce(false);
    instance.handleSubmit(event);
    expect(instance.setFormValidity).toHaveBeenCalled();
  });

  it('Should set the state based on changes from the form', () => {
    const event = {
      target: {
        name: 'email',
        value: 'test@testmail.com',
      },
    };
    instance.handleChange(event);
    wrapper.state().email.should.equal(event.target.value);
  });

  it('ValidateForm() should return true if there are no errors', () => {
    instance.validateForm().should.equal(true);
  });

  it('SetFormValidity() Should return true if there are no errors', () => {
    instance.setFormValidity(state.errors).should.equal(true);
  });

  it('Should return false if there are errors', () => {
    instance.setFormValidity({ email: 'mock error' }).should.equal(false);
    wrapper.state().isFormValid.should.equal(false);
  });

  it('toggleVisibility() should make password visible', () => {
    instance.toggleVisibility();
    wrapper.state().showPassword.should.equal(true);
  });
  it('should set window location to backend url for google', () => {
    const event = {
      target: {
        getAttribute: () => 'google',
      },
    };
    instance.handleSocialSignin(event);
  });
  it('should set window location to backend url for facebook', () => {
    const event = {
      target: {
        getAttribute: () => 'facebook',
      },
    };
    instance.handleSocialSignin(event);
  });
  it('on mounting socialLogin should be false', () => {
    instance.componentDidMount = jest.fn();
    instance.componentDidMount();
    const localStorage = {
      getItem: jest.fn().mockReturnValueOnce(false),
    };
    const socialLogin = localStorage.getItem('socialLogin');
    expect(socialLogin).toBe(false);
  });
  it('should mock componentDidMount', () => {
    const window = {
      location: {
        search: jest.fn().mockReturnValueOnce(false),
      },
    };
    const URLSearchParams = jest.fn();
    const searchParams = new URLSearchParams(window.location.search);
    searchParams.get = jest.fn().mockReturnValueOnce(false);
    const user = searchParams.get('user');
    expect(user).toBe(false);
  });
});
