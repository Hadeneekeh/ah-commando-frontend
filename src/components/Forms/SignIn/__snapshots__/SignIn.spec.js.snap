// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test social media sign in should click the google signin button 1`] = `
<Modal
  close={[Function]}
  data-test="signup-form"
  open={true}
>
  <div
    className="form"
  >
    <h3
      className="form-header"
    >
      Sign in
    </h3>
    <p
      className="form-text"
    >
      Sign in to experience authorâ€™s haven great community of creative minds. Follow your favourite authors and interact with the articles you love
    </p>
    <form
      data-test="signin-form"
      style={
        Object {
          "display": "flex",
          "flexDirection": "column",
          "justifyContent": "center",
        }
      }
    >
      <Input
        error=""
        handleChange={[Function]}
        handleToggle={null}
        id="email"
        label="Email"
        name="email"
        placeholder="john.doe@foo.bar"
        style={Object {}}
        togglable={false}
        type="email"
        value=""
        visible={false}
      />
      <Input
        error=""
        handleChange={[Function]}
        handleToggle={[Function]}
        id="password"
        label="Password"
        name="password"
        placeholder="**********"
        style={Object {}}
        togglable={true}
        type="password"
        value=""
        visible={false}
      />
      <Button
        className=""
        datatest="loginButton"
        disabled={false}
        handleClick={[Function]}
        id="signin"
        label="Sign in"
        style={
          Object {
            "backgroundColor": "#000",
            "borderRadius": "0",
            "color": "#ffc700",
            "height": "45px",
            "marginLeft": "5%",
            "width": "90%",
          }
        }
        type="submit"
      />
    </form>
    <div
      className="alternative-login"
    >
      <p>
        Or create an account Using:
      </p>
      <div
        className="social-login"
      >
        <span
          id="google"
          name="google"
          onClick={[Function]}
        >
          <i
            className="fab fa-google fa-lg"
            name="google"
            onClick={[Function]}
            style={
              Object {
                "color": "red",
              }
            }
          />
          Google
        </span>
        <span
          id="facebook"
          name="facebook"
          onClick={[Function]}
        >
          <i
            className="fab fa-facebook fa-lg"
            name="facebook"
            onClick={[Function]}
            style={
              Object {
                "color": "blue",
              }
            }
          />
          Facebook
        </span>
      </div>
    </div>
    <div
      className="switch-context"
    >
      <div>
        No account?
         
        <p
          id="sc-sn"
          onClick={[Function]}
        >
          Sign up
        </p>
      </div>
    </div>
    <div
      className="forgot-password"
    >
      <p
        className="fg-ps"
        onClick={[Function]}
      >
        Forgot password?
      </p>
    </div>
  </div>
</Modal>
`;
